nvflare/fuel/hci/zip_utils.py

```diff
@@ -165,25 +157,38 @@ def convert_legacy_zip(zip_data: bytes) -> bytes:
         The converted zip data
     """
 
+    meta: Optional[dict] = None
     reader = io.BytesIO(zip_data)
     with ZipFile(reader, "r") as in_zip:
         info_list = in_zip.infolist()
         folder_name = info_list[0].filename.split("/")[0]
         meta_path = _path_join(folder_name, META_NAME)
         if next((info for info in info_list if info.filename == meta_path), None):
             # Already in job layout
-            return zip_data
+            meta_data = in_zip.read(meta_path)
+            meta = json.loads(meta_data)
+            if JobMetaKey.JOB_FOLDER_NAME not in meta:
+                meta[JobMetaKey.JOB_FOLDER_NAME.value] = folder_name
+            else:
+                return zip_data
 
         writer = io.BytesIO()
         with ZipFile(writer, "w") as out_zip:
-            out_zip.writestr(meta_path, META_DATA.format(folder_name, folder_name))
-            # Push everything else to a sub folder with the same name:
-            # hello-pt/README.md -> hello-pt/hello-pt/README.md
-            for info in info_list:
-                name = info.filename
-                content = in_zip.read(name)
-                path = folder_name + "/" + name
-                info.filename = path
-                out_zip.writestr(info, content)
+            if meta:
+                out_zip.writestr(meta_path, json.dumps(meta))
+                out_zip.comment = in_zip.comment  # preserve the comment
+                for item in in_zip.infolist():
+                    if item.filename != meta_path:
+                        out_zip.writestr(item, in_zip.read(item.filename))
+            else:
+                out_zip.writestr(meta_path, META_DATA.format(folder_name, folder_name))
```

Comment:
this line does not guarantee we have `JobMetaKey.JOB_FOLDER_NAME` in meta.json

Reply:
since I check if the job_folder_name exist or not, the meta will be None or not None. 
if the name doesn't exists, the meta.json exists, the meta will be not None. so it will flow to the else part.

Reply:
2. When meta.json does not exist, this `meta` will be an empty dict so it enters this else statement.
    In L184 you are trying to generate a meta.json.
    I am saying this meta.json you generated here does NOT have `JobMetaKey.JOB_FOLDER_NAME`
    it only has "name", you can see this in L24 of this file
    So you need to change that `META_DATA` string to have `JobMetaKey.JOB_FOLDER_NAME`
    otherwise deploy will fail.